opennlp-morfologik-addon/src/main/java/opennlp/morfologik/lemmatizer/MorfologikLemmatizer.java

```diff
@@ -47,7 +47,7 @@ public MorfologikLemmatizer(Dictionary dictionary) throws IllegalArgumentExcepti
     dictLookup = new DictionaryLookup(dictionary);
   }
 
-  private List<String> lemmatize(String word, String postag) {
+  private synchronized List<String> lemmatize(String word, String postag) {
```

Comment:
I wonder if that's the best option we have. We could also investigate locking only on `dictLookup` (which is the only data structure in this method that is outside of its context I think?). Or using a different structure for `dictLookup` if possible. But I haven't looked at the rest of the code in a long time.

Reply:
Yes, I tried locking only around `dictLookup`, but the List returned by `lookup` method (morfologik code) returns a reference for an object that is created only once, maybe for performance reasons, I don't know :(

Reply:
I tend to think predictability is more important than impact to performance. A wrong answer isn't helpful no matter how fast it gets returned. So I would favor the `synchronized` as long as we're sure it's the best implementation choice.

Reply:
An alternative to `synchronized` would be to re-create and dispose the `DictionaryLookup`, which is " is cheap to create and dispose (so it makes no sense to cache)" according to https://github.com/morfologik/morfologik-stemming/issues/69, i.e. moving to 

```java
List<WordData> dictMap = new DictionaryLookup(dictionary).lookup(word.toLowerCase());
```

and just store the (thread-safe) dictionary in the `MorfologikLemmatizer` instead of the not thread-safe `DictionaryLookup`. 

That would avoid the synchronization cost. Wdyt?

Reply:
>That would avoid the synchronization cost. Wdyt?

That sounds like a good option to me. Avoids the `synchronized`, shouldn't add too much overhead, and the lib developers suggest it too. +1, thanks @rzo1 !

Reply:
Here is the alternative PR on the basis of @avanco work: https://github.com/apache/opennlp/pull/436